core/src/main/scala/no/nrk/bigquery/JobLabels.scala

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 NRK
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+package no.nrk.bigquery
+
+import cats.data.NonEmptyChain
+
+import scala.annotation.implicitNotFound
+import scala.collection.immutable.SortedMap
+
+@implicitNotFound("No joblabels instance was found, make sure you have an implicit or given instance")
+final case class JobLabels private (value: SortedMap[String, String]) {
+  def ++(other: JobLabels): JobLabels =
+    JobLabels(other.value ++ this.value)
+}
+
+object JobLabels {
+  import cats.syntax.all.*
+
+  val empty = JobLabels(SortedMap.empty[String, String])
+
+  def validated(values: SortedMap[String, String]): Either[NonEmptyChain[String], JobLabels] = {
+    def check(key: String)(op: String => Boolean, msg: String) =
+      if (op(key)) msg.invalidNec else key.validNec
+
+    def checkLength(v: String, mode: String) = check(v)(_.length > 63, show"label $mode $v is longer than 63 chars")
+    def checkChars(v: String, mode: String) = check(v)(
+      !_.forall(c => c.isLower || c.isDigit || c == '-' || c == '_'),
+      show"label $mode $v can contain only lowercase letters, numeric characters, underscores, and dashes"
+    )
+
+    values.toList
+      .traverse { case (key, value) =>
+        val checkKey =
+          (
+            check(key)(_.isEmpty, show"label key $key is empty"),
+            checkLength(key, "key"),
+            checkChars(key, "key"),
+            check(key)(!_(0).isLower, show"label key $key must start with lowercase letter or international character"))
+            .mapN((_, _, _, _) => key)
+
+        val checkValue = (checkLength(value, "value"), checkChars(value, "value")).mapN((_, _) => value)
+
+        (checkKey, checkValue).tupled
+      }
+      .map(list => JobLabels(SortedMap(list: _*)))
+      .toEither
+  }
+
+  /** truncates string to 63 chars. Replaces all non-word chars with _ Users should make sure this cannot result in an
+    * empty string, as that will fail validation,
+    *
+    * @param value
+    *   a non-empty string
+    * @return
+    *   a truncated or empty string.
+    */
+  def safeString(value: String): String =
+    value.trim.substring(0, Math.min(63, value.length)).toLowerCase().replaceAll("\\W", "_")
+
+  def apply(params: (String, String)*): Either[NonEmptyChain[String], JobLabels] = validated(SortedMap(params: _*))
+
+  def unsafeFrom(params: (String, String)*): JobLabels =
+    validated(SortedMap(params: _*))
+      .fold(messages => throw new IllegalArgumentException(messages.toList.mkString("\n")), identity)
```

Comment:
Should we consider using `literally` here for the labels (keys and value). My experience using labels is that it's easy to get wrong and often detected to late so it had been nice if we could catch this as easy as possible.

Reply:
`literally` requires a datatype, so I think a custom macro would be needed here.